helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/RandomRecordGenerator.java

```diff
@@ -98,12 +99,36 @@ private Object newRandomGeneric(Schema of, RecordGenerationConfig config) {
         }
         return map;
       case UNION:
-        List<Schema> branches = of.getTypes();
-        index = random.nextInt(branches.size());
-        Schema branch = branches.get(index);
+        List<Schema> acceptableBranches = narrowDownUnionBranches(of, of.getTypes(), config);
+        index = random.nextInt(acceptableBranches.size());
+        Schema branch = acceptableBranches.get(index);
         return newRandomGeneric(branch, config);
       default:
         throw new UnsupportedOperationException("unhandled: " + of.getType());
     }
   }
+
+  /**
+   * given a union schema, narrow down "acceptable" union branches that match the given generation config
+   * @param unionSchema union schema
+   * @param branches branches of the union schema
+   * @param config config to determine which branches are "acceptable"
+   * @return acceptable branches. throws an exception if no branches meet criteria
+   */
+  private List<Schema> narrowDownUnionBranches(Schema unionSchema, List<Schema> branches, RecordGenerationConfig config) {
+    List<Schema> results = new ArrayList<>(branches.size());
+    for (Schema branch : branches) {
+      if (isAcceptableUnionBranch(unionSchema, branch, config)) {
+        results.add(branch);
+      }
+    }
+    if (results.isEmpty()) {
+      throw new IllegalStateException("no acceptable union branches out of original " + unionSchema);
+    }
+    return results;
+  }
+
+  private boolean isAcceptableUnionBranch(Schema unionSchema, Schema proposedBranch, RecordGenerationConfig config) {
+    return !config.avoidNulls() || proposedBranch.getType() != Schema.Type.NULL;
```

Comment:
Nitpick... the double negative OR can refactored into the (IMHO) more readable form of a double positive AND.

Reply:
inverting this condition leads to !(avoidNulls && branch.type==null). jury's out on readability here, but will do

Reply:
As you wish. It's just a personal opinion. Totally fine to disregard